File: PrebidMobile/Objc/PrebidMobileRendering/ORTB/PBMORTBDeal.m
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "PBMORTBDeal.h"
#import "PBMORTBAbstract+Protected.h"

@implementation PBMORTBDeal

- (nonnull instancetype)init {
    if (!(self = [super init])) {
        return nil;
    }
    _bidfloor = @(0.0);
    _bidfloorcur = @"USD";
    _wseat = @[];
    _wadomain = @[];
    
    return self;
}
- (void)setWseat:(NSArray<NSString *> *)wseat {
    _wseat = wseat ? [NSArray arrayWithArray:wseat] : @[];
}
- (void)setWadomain:(NSArray<NSString *> *)wadomain {
    _wadomain = wadomain ? [NSArray arrayWithArray:wadomain] : @[];
}

- (nonnull PBMJsonDictionary *)toJsonDictionary {
    PBMMutableJsonDictionary *ret = [PBMMutableJsonDictionary new];
    
    ret[@"id"] = self.id;
    ret[@"bidfloor"] = self.bidfloor;
    ret[@"bidfloorcur"] = self.bidfloorcur;
    ret[@"at"] = self.at;
    ret[@"wseat"] = self.wseat;
    ret[@"wadomain"] = self.wadomain;
    
    ret = [ret pbmCopyWithoutEmptyVals];
    
    return ret;
}

- (instancetype)initWithJsonDictionary:(nonnull PBMJsonDictionary *)jsonDictionary {
    if (!(self = [self init])) {
        return nil;
    }
    _id = jsonDictionary[@"id"];
    _bidfloor = jsonDictionary[@"bidfloor"];
    _bidfloorcur = jsonDictionary[@"bidfloorcur"];
    _at = jsonDictionary[@"at"];
    _wseat = jsonDictionary[@"wseat"];
    _wadomain = jsonDictionary[@"wadomain"];
    
    return self;
}

@end

File: PrebidMobile/Swift/PrebidMobileRendering/ORTB/Request/PBMORTBDeal.swift
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import Foundation

@objc public class PBMORTBDeal: NSObject, PBMJsonCodable {

    // MARK: - Properties

    @objc public var id: String?
    @objc public var bidfloor: NSNumber = 0.0
    @objc public var bidfloorcur: String = "USD"
    @objc public var at: NSNumber?
    @objc public var wseat: [String] = []
    @objc public var wadomain: [String] = []

    // MARK: - Init

    public override init() {
        super.init()
    }

    @objc(initWithJsonDictionary:)
    public required init(jsonDictionary: [String: Any]) {
        super.init()
        let json = JSONObject<Key>(jsonDictionary)
        id          = json[.id]
        bidfloor    = json[.bidfloor] ?? 0.0
        bidfloorcur = json[.bidfloorcur] ?? "USD"
        at          = json[.at]
        wseat       = (json[.wseat] as [String]?) ?? []
        wadomain    = (json[.wadomain] as [String]?) ?? []
    }

    // MARK: - PBMJsonEncodable

    @objc(toJsonDictionary)
    public var jsonDictionary: [String: Any] {
        var json = JSONObject<Key>()
        json[.id]          = id
        json[.bidfloor]    = bidfloor
        json[.bidfloorcur] = bidfloorcur
        json[.at]          = at
        json[.wseat]       = wseat
        json[.wadomain]    = wadomain
        return json.dict
    }

    // MARK: - Keys

    private enum Key: String {
        case id, bidfloor, bidfloorcur, at, wseat, wadomain
    }
}
